contracts/contracts/strategies/sonic/SonicSwapXAMOStrategy.sol

```diff
@@ -0,0 +1,710 @@
+        view
+        override
+        returns (uint256 balance)
+    {
+        require(_asset == ws, "Unsupported asset");
+
+        // wS balance needed here for the balance check that happens from vault during depositing.
+        balance = IERC20(ws).balanceOf(address(this));
+
+        // This assumes 1 gauge LP token = 1 pool LP token
+        uint256 lpTokens = IGauge(gauge).balanceOf(address(this));
+        if (lpTokens == 0) return balance;
+
+        // Add the strategy’s share of the wS and OS tokens in the SwapX pool if the pool was balanced.
+        balance += _lpValue(lpTokens);
+    }
+
+    /**
+     * @notice Returns bool indicating whether asset is supported by strategy
+     * @param _asset Address of the asset
+     */
+    function supportsAsset(address _asset) public view override returns (bool) {
+        return _asset == ws;
+    }
+
+    /**
+     * @notice Collect accumulated SWPx (and other) rewards and send to the Harvester.
+     */
+    function collectRewardTokens()
+        external
+        override
+        onlyHarvester
+        nonReentrant
+    {
+        // Collect SWPx rewards from the gauge
+        IGauge(gauge).getReward();
+
+        _collectRewardTokens();
+    }
+
+    /***************************************
+        Internal SwapX Pool and Gauge Functions
+    ****************************************/
+
+    /**
+     * @dev Calculate the required amount of OS to mint based on the wS amount.
+     * This ensures the proportion of OS tokens being added to the pool matches the proportion of wS tokens.
+     * For example, if the added wS tokens is 10% of existing wS tokens in the pool,
+     * then the OS tokens being added should also be 10% of the OS tokens in the pool.
+     * @param _wsAmount Amount of Wrapped S (wS) to be added to the pool.
+     * @return osAmount Amount of OS to be minted and added to the pool.
+     */
+    function _calcTokensToMint(uint256 _wsAmount)
+        internal
+        view
+        returns (uint256 osAmount)
+    {
+        (uint256 wsReserves, uint256 osReserves, ) = IPair(pool).getReserves();
+        require(wsReserves > 0, "Empty pool");
+
+        // OS to add = (wS being added * OS in pool) / wS in pool
+        osAmount = (_wsAmount * osReserves) / wsReserves;
+    }
+
+    /**
+     * @dev Calculate how much pool LP tokens to burn to get the required amount of wS tokens back
+     * from the pool.
+     * @param _wsAmount Amount of Wrapped S (wS) to be removed from the pool.
+     * @return lpTokens Amount of SwapX pool LP tokens to burn.
+     */
+    function _calcTokensToBurn(uint256 _wsAmount)
+        internal
+        view
+        returns (uint256 lpTokens)
+    {
+        /* The rate between coins in the pool determines the rate at which pool returns
+         * tokens when doing balanced removal (remove_liquidity call). And by knowing how much wS
+         * we want we can determine how much of OS we receive by removing liquidity.
```

Comment:
not sure should this comment say: 
"... we want we can determine how much LP we need to burn in order to receive the desired amount of wS
""

Reply:
This was copied from the Curve AMO which is a little different in Curve supports unbalanced removals.

I've changed it to this
```
  /* The SwapX pool proportionally returns the reserve tokens when removing liquidity.
   * First, calculate the proportion of required wS tokens against the pools wS reserves.
   * That same proportion is used to calculate the required amount of pool LP tokens.
   * For example, if the required wS tokens is 10% of the pool's wS reserves,
   * then 10% of the pool's LP supply needs to be burned.
```